c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -1210,6 +1212,1279 @@ namespace NativeStreamSource {
+      if (!then_handler) return false;
+
+      // Store the chunk value on the then_handler function for later retrieval.
+      js::SetFunctionNativeReserved(then_handler, 1, chunk);
+
+      RootedObject result(cx);
+      result = JS::CallOriginalPromiseThen(cx, changePromise, then_handler, nullptr);
+      if (!result) {
+        return false;
+      }
+
+      args.rval().setObject(*result);
+      return true;
+    }
+
+    // 2.  Let controller be stream.[controller].
+    RootedObject controller(cx, TransformStream::controller(stream));
+
+    // 4.  Return ! [TransformStreamDefaultControllerPerformTransform](controller, chunk).
+    RootedObject transformPromise(cx);
+    transformPromise = TransformStreamDefaultController::PerformTransform(cx, controller, chunk);
+    if (!transformPromise) {
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+    }
+
+    args.rval().setObject(*transformPromise);
+    return true;
+  }
+
+  bool DefaultSinkAbortAlgorithm(JSContext* cx, CallArgs args,
+                                 HandleObject writableController, HandleObject stream,
+                                 HandleValue reason)
+  {
+    MOZ_ASSERT(is_instance(stream));
+
+    // 1.  Perform ! [TransformStreamError](stream, reason).
+    if (!Error(cx, stream, reason)) {
+      return false;
+    }
+
+    // 2.  Return [a promise resolved with] undefined.
+    RootedObject promise(cx, JS::CallOriginalPromiseResolve(cx, JS::UndefinedHandleValue));
+    if (!promise) {
+      return false;
+    }
+
+    args.rval().setObject(*promise);
+    return true;
+  }
+
+  /**
+   * Steps 5.1.1-2 of DefaultSinkCloseAlgorithm.
+   */
+  bool default_sink_close_algo_then_handler(JSContext* cx, HandleObject stream, CallArgs args) {
+    // 5.1.1.  If readable.[state] is "`errored`", throw readable.[storedError].
+    RootedObject then_handler(cx, &args.callee());
+    RootedObject readable(cx, &js::GetFunctionNativeReserved(then_handler, 1).toObject());
+    bool is_errored;
+    if (!JS::ReadableStreamIsErrored(cx, readable, &is_errored)) {
+      return false;
+    }
+
+    if (is_errored) {
+      RootedValue storedError(cx, JS::ReadableStreamGetStoredError(cx, readable));
+      JS_SetPendingException(cx, storedError);
+      return false;
+    }
+
+    // 5.1.2.  Perform ! [ReadableStreamDefaultControllerClose](readable.[controller]).
+    RootedObject readableController(cx, JS::ReadableStreamGetController(cx, readable));
+    // Note: in https://github.com/whatwg/streams/pull/1029, the spec was changed to
+    // make ReadableStreamDefaultControllerClose (and -Enqueue) return early if
+    // ReadableStreamDefaultControllerCanCloseOrEnqueue is false. SpiderMonkey hasn't been
+    // updated accordingly, so it'll throw an exception instead.
+    // To avoid that, we do the check explicitly. While that also throws an exception, we can
+    // just clear it and move on.
```

Comment:
Same comment here as above: we should note what will happen when SM is updated, so that this code is future-proof.